### PR TITLE
GraphQL scalars: fixes and tests

### LINKED ISSUE #206: Support Duration GraphQL scalar types
We should support Duration CQL data type in GraphQL.

### LINKED ISSUE #261: GraphQL: Create integration tests with every supported scalar type
We should cover all supported scalars in the integration test suite.

### LINKED ISSUE #268: GraphQL: fix handling of DATE type
`CustomScalar.DATE` uses the following to parse GraphQL values:
```java
o -> java.sql.Date.valueOf(o.toString())
```

This is not going to work. That value is later passed to `QueryBuilder.literal()`, which uses the default Java driver mappings. DATE maps to `java.time.LocalDate`, not `java.sql.Date`.

### LINKED ISSUE #272: Use graphql-java builtin scalars for CQL scalars
Use `GraphQLShort`, `GraphQLLong`, `GraphQLByte`, `GraphQLBigDecimal` and `GraphQLBigInteger` to represent their CQL counterparts.

See also #261 .

### LINKED ISSUE #304: GraphQL Scalars: Blob encoding to string is broken
It's returned as a byte array instead of a base64 string.

### PR DESCRIPTION
**Pull request by jorgebay**
Unit and integration tests for each scalar type. Contains various fixes.

Resolves #261, resolves #206, resolves #304, resolves #268 and resolves #272.

### CODE REVIEW

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ persistence-api/src/main/java/io/stargate/db/datastore/query/WhereCondition.java  19
+ persistence-api/src/main/java/io/stargate/db/datastore/ResultSet.java  46
+ persistence-api/src/main/java/io/stargate/db/Result.java  78
+ persistence-api/src/main/java/io/stargate/db/BatchType.java  90
+ persistence-api/src/main/java/io/stargate/db/schema/Table.java  19
+ persistence-api/src/main/java/io/stargate/db/schema/Column.java  10
+ persistence-api/src/main/java/io/stargate/db/schema/Keyspace.java  11
+ persistence-api/src/main/java/io/stargate/db/datastore/query/Where.java  28
+ persistence-api/src/main/java/io/stargate/db/datastore/query/Parameter.java  33
+ persistence-api/src/main/java/io/stargate/db/datastore/query/QueryBuilderImpl.java  42
+ persistence-api/src/main/java/io/stargate/db/schema/SecondaryIndex.java  31
+ persistence-api/src/main/java/io/stargate/db/EventListener.java  3
+ persistence-api/src/main/java/io/stargate/db/schema/Schema.java  9
+ persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedDataStore.java  77
+ persistence-api/src/main/java/io/stargate/db/datastore/query/ColumnOrder.java  33
+ persistence-api/src/main/java/io/stargate/db/Persistence.java  80
+ persistence-api/src/main/java/io/stargate/db/datastore/query/MixinPreparedStatement.java  85
+ persistence-api/src/main/java/io/stargate/db/datastore/DataStore.java  93
+ persistence-api/src/main/java/io/stargate/db/datastore/query/Value.java  62
+ persistence-api/src/main/java/io/stargate/db/schema/UserDefinedType.java  3
+ persistence-api/src/main/java/io/stargate/db/ClientInfo.java  56
+ persistence-api/src/main/java/io/stargate/db/DriverInfo.java  50
+ persistence-api/src/main/java/io/stargate/db/schema/ParameterizedType.java  5
+ persistence-api/src/main/java/io/stargate/db/BoundStatement.java  80
+ persistence-api/src/main/java/io/stargate/db/Statement.java  100
+ persistence-api/src/main/java/io/stargate/db/schema/CollectionIndexingType.java  100
+ persistence-api/src/main/java/io/stargate/db/datastore/PreparedStatement.java  80
+ graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java  39
+ persistence-api/src/main/java/io/stargate/db/schema/ColumnUtils.java  3
+ persistence-api/src/main/java/io/stargate/db/datastore/ArrayListBackedRow.java  67
+ persistence-api/src/main/java/io/stargate/db/datastore/query/NWhere.java  46
+ persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedPreparedStatement.java  57
+ persistence-api/src/main/java/io/stargate/db/datastore/query/AndWhere.java  20
+ persistence-api/src/main/java/io/stargate/db/Batch.java  100
+ persistence-api/src/main/java/io/stargate/db/AuthenticatedUser.java  100
+ persistence-api/src/main/java/io/stargate/db/SimpleStatement.java  88
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DataTypeMapping.java  15
+ persistence-api/src/main/java/io/stargate/db/Parameters.java  73
+ persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedResultSet.java  63
+ persistence-api/src/main/java/io/stargate/db/schema/AbstractTable.java  69
+ graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java  1
         

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6390757&bid=20727510)

**Review by olim7t — APPROVED** [review]
Looks great, especially the tests.

Minor remarks except for the one about literal vs value parsing, but maybe that's a broader issue to be addressed as a follow-up.

**Review by dougwettlaufer — APPROVED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java [hunk 1]
```diff
@@ -225,10 +239,15 @@ public static Object literalValueToObject(Value value) {
       return ((StringValue) value).getValue();
     }
     if (value instanceof IntValue) {
-      return ((IntValue) value).getValue();
+      // Int scalar type represents a signed 32‐bit numeric non‐fractional value.
+      // But, the JSON Number is a 64-bit double than can represents ints up to 2^53
+      // So we choose to represent it as a long
+      return ((IntValue) value).getValue().longValue();
```
[the remark this answers is not in the document]
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java:245]: To answer my own question: `Coercing.parseValue()` is used when a query is parameterized. For example via the Playground UI:
```
query foo($u: Uuid!) {
    foo(value: { u: $u }) { values { u } }
}
```
And in the "Query Variables" pane:
```
{"u": "a35c0ac1-1574-11eb-8269-333197b785d7"}
```
The variables must be in JSON format, they are sent as a separate field in the JSON payload (only the contents of the "query" field are in GraphQL format).

The server-side deserialization happens in graphql-java-servlet, more precisely [here](https://github.com/graphql-java-kickstart/graphql-java-servlet/blob/master/graphql-java-kickstart/src/main/java/graphql/kickstart/execution/GraphQLObjectMapper.java#L175-L176). It's done with a Jackson object mapper that gets configured [here](https://github.com/graphql-java-kickstart/graphql-java-servlet/blob/master/graphql-java-kickstart/src/main/java/graphql/kickstart/execution/error/DefaultGraphQLServletObjectMapperConfigurer.java) (we don't override the config so it's pretty standard). Bottom line: the types that we should be prepared to handle in `parseValue()` are the default Jackson mappings: Integer, Long, BigInteger, Double (or BigDecimal after #325), Boolean, null, Map<String,Object>, List of any of the above.

I still think the code could be improved, but that's off topic. I'll create another ticket.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java [hunk 2]
```diff
@@ -49,17 +51,15 @@
   UUID(
       "Uuid",
       "The `Uuid` scalar type represents a CQL uuid as a string.",
-      o -> java.util.UUID.fromString(String.valueOf(o)),
-      Object::toString),
+      o -> java.util.UUID.fromString(String.valueOf(o))),
```
[the remark this answers is not in the document]
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java:54]: In hindsight, the codec's null handling is not really appropriate here, GraphQL has its own null type, we should use that, not a string containing the word "null ".

There's still some value in reusing the codecs, if we want to support the same rules as CQL the parsing is not trivial, see `TimestampCodec` for example.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6390757&bid=20727510)

File: graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java:245]: Yeah, as we are using GraphQL over JSON, GraphQL `Int` will be passed along as JSON `Number` (IEEE double) literals. IEEE double can represent _exact_ integers up to 2^53.

I'm not sure why graphql-java uses `BigInteger` to represent the value, probably it was not a good decision.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/CustomScalar.java:54]: Good call about leveraging existing codecs, I've filed #326 for it.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java [hunk 3]
```diff
@@ -40,6 +40,12 @@ private ColumnType normalize(ColumnType type) {
       type = ImmutableListType.builder().addAllParameters(type.parameters()).build();
     }
 
+    if (type == Type.Varchar) {
+      // Currently field type cache does not tolerate 2 different column types (Text and Varchar)
+      // using the same graphql type, workaround by using a single column type
+      type = Type.Text;
+    }
```
[the remark this answers is not in the document]
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java:51]: Makes sense, I've added some general doc comments to `normalize()` method.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6390757&bid=20727510)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6390757&bid=20727510)

**Pull request merged**